src/org/sosy_lab/java_smt/solvers/mathsat5/Mathsat5IntegerFormulaManager.java

```diff
@@ -69,6 +77,21 @@ public Long divide(Long pNumber1, Long pNumber2) {
         msat_make_floor(mathsatEnv, div));
   }
 
+  @Override
+  protected Long modulo(Long pNumber1, Long pNumber2) {
+    // The modulo can be calculated by the formula:
+    //   remainder = dividend - floor(dividend/divisor)*divisor
+    // However, this will fail if the divisor is zero as SMTLIB then allows the solver to return
+    // any value for the remainder. We solve this by first checking the divisor and returning an
+    // UF symbol "modZeroUF(dividend)" if it is zero. Otherwise, the formula is used to calculate
+    // the remainder.
+    return msat_make_term_ite(
+        mathsatEnv,
+        msat_make_equal(mathsatEnv, pNumber2, msat_make_int_number(mathsatEnv, 0)),
+        getFormulaCreator().callFunctionImpl(modZeroUF, ImmutableList.of(pNumber1)),
+        subtract(pNumber1, multiply(divide(pNumber1, pNumber2), pNumber2)));
+  }
+
```

Comment:
Be aware that UFs return the same value for the same parameters.
If SMTLIB allows "any return value", then the return value could be random and return a different result at second call.
From technical point, returning the same result on second call would be fully ok for me. From SMTLIB point, please check what the standard and other solvers return for `x/0 == x/0`. Is this always satisfied?

Reply:
Valid point. Lets approve it.